### Experiment 1: Breadth-First Search (BFS) for Maze Navigation
**Problem Statement:**

Consider a maze represented as a graph, where each node corresponds to a location in the maze and edges represent valid paths between locations. An agent is placed at a start node and must reach a goal node by traversing the maze.

**Task:**
1. Formulate the maze navigation problem as a search problem by clearly defining:
    * State space
    * Initial state
    * Actions
    * Transition model
    * Goal test
    * Path cost
2. Implement the Breadth-First Search (BFS) algorithm to find the shortest path from the start node to the goal node.
3. Print the path returned by BFS.
4. Count and display the number of nodes expanded during the search.
   
**Constraints:**
* The maze is unweighted.
* The environment is fully observable, deterministic, static, discrete, and single-agent.
* BFS must be implemented without using any built-in graph search libraries.

In [ ]:
class Queue:
    def __init__(self, size=100):
        self.size = size
        self.queue = []
    def enqueue(self, element):
        if len(self.queue) >= self.size:
            return False
        self.queue.append(element)
        return True
    def dequeue(self):
        if len(self.queue) <= 0:
            return None
        element = self.queue[0]
        del self.queue[0]
        return element
    def isEmpty(self):
        return len(self.queue) <= 0

In [ ]:
def BFS(graph, start, end):
    queue = Queue()
    queue.enqueue(start)
    visited = set()
    parent = {}
    parent[start] = None

    while not queue.isEmpty():
        node = queue.dequeue()
        visited.add(node)

        if node == end:
            return reconstruct_path(parent, start, end)

        for neighbor in graph[node]:
            if neighbor not in visited and neighbor not in parent:
                parent[neighbor] = node
                queue.enqueue(neighbor)

    return None

def reconstruct_path(parent, start, end):
    path = []
    current = end

    while current is not None:
        path.append(current)
        current = parent[current]

    path.reverse()
    return path

In [ ]:
maze = {
    'A': ['B', 'C'],
    'B': ['A', 'D', 'E'],
    'C': ['A', 'F'],
    'D': ['B'],
    'E': ['B', 'F'],
    'F': ['C', 'G', 'E'],
    'G': [],
}

path = BFS(maze, 'A', 'G')
print("Path: " + " -> ".join(path))

Path: A -> C -> F -> G


### Experiment 2: Depth-First Search (DFS) for Graph Traversal
**Problem Statement:**

Given a connected, finite graph represented using an adjacency list, design and implement the Depth-First Search (DFS) algorithm to find a path from a specified start node to a goal node.

**Task Requirements:**
1. Formulate the graph traversal problem by identifying:
    * State representation
    * Initial state
    * Goal state
    * Actions
2. Implement a recursive DFS algorithm that:
    * Explores one branch of the search tree as deeply as possible before backtracking
    * Avoids revisiting nodes already present in the current path
3. Return and display:
    * A valid path from the start node to the goal node (if it exists)

**Constraints:**
* The graph is unweighted.
* DFS must be implemented without using any built-in graph traversal libraries.
* The algorithm is not required to return the shortest path.

**Expected Outcome:**

The algorithm should successfully find a path from the start node to the goal node by exploring depth-wise, though the path may not be optimal.

In [ ]:
class Stack:
    def __init__(self, size=100):
        self.size = size
        self.queue = []
    def push(self, element):
        if len(self.queue) >= self.size:
            return False
        self.queue.append(element)
        return True
    def pop(self):
        if len(self.queue) <= 0:
            return None
        element = self.queue.pop()
        return element
    def isEmpty(self):
        return len(self.queue) <= 0

In [ ]:
def DFS(graph, start, end):
    stack = Stack()
    stack.push(start)
    visited = set()
    parent = {}
    parent[start] = None

    while not stack.isEmpty():
        node = stack.pop()
        visited.add(node)

        if node == end: return reconstruct_path(parent, start, end)

        for neighbor in graph[node]:
            if neighbor not in visited and neighbor not in parent:
                parent[neighbor] = node
                stack.push(neighbor)

    return None

In [ ]:
maze = {
    'A': ['B', 'C'],
    'B': ['A', 'D', 'E'],
    'C': ['A', 'F'],
    'D': ['B'],
    'E': ['B', 'F'],
    'F': ['C', 'G', 'E'],
    'G': [],
}

path = DFS(maze, 'A', 'G')
print("Path: " + " -> ".join(path))

Path: A -> C -> F -> G


### Experiment 3: Solving the 8-Puzzle Problem Using Breadth-First Search (BFS)

**Problem Statement:**

The 8-puzzle problem consists of a 3×3 grid containing eight numbered tiles and one blank space. A tile adjacent to the blank space can be moved into the blank position. Given an initial configuration and a goal configuration, design and implement the Breadth-First Search (BFS) algorithm to find a sequence of moves that transforms the initial state into the goal state.

**Task Requirements:**
1. Formulate the 8-puzzle problem by identifying:
    * State representation
    * Initial state
    * Goal state
    * Valid actions (Up, Down, Left, Right)
    * Transition model
2. Implement the Breadth-First Search (BFS) algorithm that:
    * Explores the state space level by level
    * Guarantees the shortest path to the goal state
    * Avoids revisiting previously explored states
3. Return and display:
    * The sequence of states from the initial state to the goal state
    * The total number of states expanded during the search

**Constraints:**
* BFS must be used to guarantee the shortest solution.
* Repeated states must be avoided using an explored set.
* Assume that the given initial state is solvable.

**Expected Outcome:**

The algorithm should successfully find the shortest sequence of moves required to transform the initial configuration into the goal configuration.


In [3]:
def find_blank(state):
    """Finds the (row, col) coordinates of the blank space (0)."""
    for r in range(3):
        for c in range(3):
            if state[r][c] == 0:
                return r, c
    return None # Should not happen in a valid 8-puzzle state

def move_blank(state, blank_pos, direction):
    """Generates a new state by moving the blank in the given direction.
    Returns None if the move is invalid."""
    r, c = blank_pos
    new_r, new_c = r, c

    if direction == 'UP':
        new_r -= 1
    elif direction == 'DOWN':
        new_r += 1
    elif direction == 'LEFT':
        new_c -= 1
    elif direction == 'RIGHT':
        new_c += 1
    else:
        return None # Invalid direction

    # Check if the new position is within bounds
    if 0 <= new_r < 3 and 0 <= new_c < 3:
        # Convert tuple of tuples to a list of lists for modification
        new_state_list = [list(row) for row in state]

        # Swap the blank with the tile at the new position
        new_state_list[r][c], new_state_list[new_r][new_c] = \
            new_state_list[new_r][new_c], new_state_list[r][c]

        # Convert back to tuple of tuples for hashability
        return tuple(tuple(row) for row in new_state_list)
    return None # Invalid move

def get_neighbors(state):
    """Returns a list of all valid neighboring states from the current state."""
    neighbors = []
    blank_r, blank_c = find_blank(state)

    directions = {'UP': (-1, 0), 'DOWN': (1, 0), 'LEFT': (0, -1), 'RIGHT': (0, 1)}

    for direction_name in ['UP', 'DOWN', 'LEFT', 'RIGHT']:
        new_state = move_blank(state, (blank_r, blank_c), direction_name)
        if new_state:
            neighbors.append(new_state)
    return neighbors

def tuple_to_string(state):
    """Converts a tuple of tuples state representation to a string for hashability."""
    return ''.join(str(tile) for row in state for tile in row)

def string_to_tuple(state_str):
    """Converts a string representation back to a tuple of tuples state."""
    # Assuming the string is a flattened 9-digit representation of the 3x3 puzzle
    return (
        tuple(int(state_str[i]) for i in range(0, 3)),
        tuple(int(state_str[i]) for i in range(3, 6)),
        tuple(int(state_str[i]) for i in range(6, 9))
    )

def bfs_8_puzzle(initial_state, goal_state):
    queue = [(initial_state, [])] # Queue stores (state, path_to_state)
    visited = set()
    expanded_nodes = 0

    # Convert states to string for hashing in visited set
    initial_state_str = tuple_to_string(initial_state)
    goal_state_str = tuple_to_string(goal_state)

    visited.add(initial_state_str)

    while queue:
        current_state, path = queue.pop(0) # Dequeue
        expanded_nodes += 1

        if tuple_to_string(current_state) == goal_state_str:
            return path + [current_state], expanded_nodes

        for neighbor in get_neighbors(current_state):
            neighbor_str = tuple_to_string(neighbor)
            if neighbor_str not in visited:
                visited.add(neighbor_str)
                queue.append((neighbor, path + [current_state]))

    return None, expanded_nodes # No solution found

In [4]:
# Define initial and goal states for the 8-puzzle
initial_state = (
    (1, 2, 3),
    (0, 4, 6),
    (7, 5, 8)
)

goal_state = (
    (1, 2, 3),
    (4, 5, 6),
    (7, 8, 0)
)

print("Initial State:")
for row in initial_state:
    print(row)

print("\nGoal State:")
for row in goal_state:
    print(row)

# Solve the puzzle using BFS
solution_path, expanded_nodes = bfs_8_puzzle(initial_state, goal_state)

if solution_path:
    print("\nSolution Found!")
    print(f"Number of expanded nodes: {expanded_nodes}")
    print("\nPath to Goal:")
    for i, state in enumerate(solution_path):
        print(f"--- Step {i} ---")
        for row in state:
            print(row)
else:
    print("\nNo solution found.")

Initial State:
(1, 2, 3)
(0, 4, 6)
(7, 5, 8)

Goal State:
(1, 2, 3)
(4, 5, 6)
(7, 8, 0)

Solution Found!
Number of expanded nodes: 17

Path to Goal:
--- Step 0 ---
(1, 2, 3)
(0, 4, 6)
(7, 5, 8)
--- Step 1 ---
(1, 2, 3)
(4, 0, 6)
(7, 5, 8)
--- Step 2 ---
(1, 2, 3)
(4, 5, 6)
(7, 0, 8)
--- Step 3 ---
(1, 2, 3)
(4, 5, 6)
(7, 8, 0)


### Experiment 4: Performance Comparison of BFS and DFS  
**Problem Statement:**

Breadth-First Search (BFS) and Depth-First Search (DFS) exhibit different performance characteristics in terms of time complexity, space complexity, and solution optimality. Design an experiment to compare BFS and DFS when applied to the same search problem.  

**Task Requirements:**  
1. Apply both BFS and DFS to a common problem instance (graph traversal or maze navigation).  
2. Instrument both algorithms to record:  
* Number of nodes expanded  
* Execution time  
* Maximum memory used (queue or recursion depth)
3. Tabulate the results obtained for BFS and DFS.  
4. Analyze and explain the observed differences.  

**Constraints:**  
* Use the same start and goal states for both algorithms.  
* Run each algorithm under identical conditions.  
* Use Python's time module for execution time measurement.  

**Expected Outcome:**

Students should observe that:  
* BFS guarantees the shortest path but uses more memory  
* DFS uses less memory but may find non-optimal paths  

**Post-Exercise Analysis Questions:**  
1. Why does BFS guarantee an optimal solution while DFS does not?  
2. In which scenarios is DFS preferred over BFS?  
3. How does the branching factor affect BFS performance?  
4. Why is BFS considered impractical for large state spaces?  

In [12]:
import time
import pandas as pd

# Helper function to reconstruct path
def reconstruct_path(parent, start, end):
    path = []
    current = end
    while current is not None:
        path.append(current)
        current = parent[current]
    path.reverse()
    return path

# BFS with tracking
def BFS_tracked(graph, start, end):
    start_time = time.time()
    queue = Queue()
    queue.enqueue(start)
    visited = set()
    parent = {}
    parent[start] = None
    expanded_nodes = 0

    while not queue.isEmpty():
        node = queue.dequeue()
        expanded_nodes += 1
        visited.add(node)

        if node == end:
            path = reconstruct_path(parent, start, end)
            end_time = time.time()
            execution_time = end_time - start_time
            return path, expanded_nodes, execution_time

        for neighbor in graph[node]:
            if neighbor not in visited and neighbor not in parent:
                parent[neighbor] = node
                queue.enqueue(neighbor)

    end_time = time.time()
    execution_time = end_time - start_time
    return None, expanded_nodes, execution_time

# DFS with tracking
def DFS_tracked(graph, start, end):
    start_time = time.time()
    stack = Stack()
    stack.push(start)
    visited = set()
    parent = {}
    parent[start] = None
    expanded_nodes = 0

    while not stack.isEmpty():
        node = stack.pop()
        expanded_nodes += 1

        if node in visited:
            continue
        visited.add(node)

        if node == end:
            path = reconstruct_path(parent, start, end)
            end_time = time.time()
            execution_time = end_time - start_time
            return path, expanded_nodes, execution_time

        for neighbor in graph[node]:
            if neighbor not in visited and neighbor not in parent:
                parent[neighbor] = node
                stack.push(neighbor)

    end_time = time.time()
    execution_time = end_time - start_time
    return None, expanded_nodes, execution_time

# Define the maze from Experiment 1
maze = {
    'A': ['B', 'C'],
    'B': ['A', 'D', 'E'],
    'C': ['A', 'F'],
    'D': ['B'],
    'E': ['B', 'F'],
    'F': ['C', 'G', 'E'],
    'G': [],
}

start_node = 'A'
goal_node = 'G'

# Run BFS_tracked
bfs_path, bfs_expanded_nodes, bfs_execution_time = BFS_tracked(maze, start_node, goal_node)

# Run DFS_tracked
dfs_path, dfs_expanded_nodes, dfs_execution_time = DFS_tracked(maze, start_node, goal_node)

# Prepare results for display
results = {
    'Algorithm': ['BFS', 'DFS'],
    'Path Found': [' -> '.join(bfs_path) if bfs_path else 'No path found', ' -> '.join(dfs_path) if dfs_path else 'No path found'],
    'Expanded Nodes': [bfs_expanded_nodes, dfs_expanded_nodes],
    'Execution Time (s)': [bfs_execution_time, dfs_execution_time]
}

df_results = pd.DataFrame(results)

print("Performance Comparison Table:")
display(df_results)


Performance Comparison Table:


,Algorithm,Path Found,Expanded Nodes,Execution Time (s)
0,BFS,A -> C -> F -> G,7,0.000022
1,DFS,A -> C -> F -> G,5,0.000012


### Analysis of BFS vs. DFS Performance

The table above shows the performance metrics for BFS and DFS on the given maze. Both algorithms successfully found a path from 'A' to 'G'.

- **BFS** found the path `A -> C -> F -> G` by expanding 7 nodes with an execution time of approximately 0.000022 seconds.
- **DFS** also found the path `A -> C -> F -> G` by expanding 5 nodes with an execution time of approximately 0.000012 seconds.

In this specific, small maze:
- **DFS was faster and expanded fewer nodes.** This is because the goal node 'G' was found relatively early along one of the deeper branches that DFS explores first (`A -> C -> F -> G`).

---

#### Post-Exercise Analysis Questions:

1.  **Why does BFS guarantee an optimal solution while DFS does not?**
    BFS explores the graph level by level, ensuring that it finds the shortest path in terms of the number of edges (cost) in an unweighted graph. It expands all nodes at depth 'd' before expanding any nodes at depth 'd+1'. Therefore, the first time it reaches the goal node, it must have found the shortest path. DFS, on the other hand, explores as deeply as possible along each branch. If the first path it finds to the goal is very long but deep, it will return that path even if a much shorter path exists at a shallower level.

2.  **In which scenarios is DFS preferred over BFS?**
    DFS is often preferred when:
    *   **Memory is a concern**: DFS's space complexity is proportional to the depth of the graph, which can be much smaller than BFS's space complexity (proportional to the number of nodes in the widest level). For very wide but not very deep graphs, DFS uses less memory.
    *   **Any solution is acceptable, not necessarily the shortest**: If the goal is simply to find a path, and optimality is not required, DFS can often find a solution more quickly by delving deep.
    *   **The graph is very deep and the solution is expected to be deep**: If the goal is likely to be found at a deep level, DFS might find it faster.

3.  **How does the branching factor affect BFS performance?**
    The branching factor (b) is the average number of children per node. For BFS, the number of nodes at depth 'd' is `b^d`. If the solution is at depth 'd', BFS might have to explore a significant portion of `b^1 + b^2 + ... + b^d` nodes. A higher branching factor dramatically increases the number of nodes to visit at each level, leading to a much larger time and space complexity, especially for deeper solutions.

4.  **Why is BFS considered impractical for large state spaces?**
    BFS is impractical for large state spaces primarily due to its **memory requirements**. In large graphs with a high branching factor, the number of nodes at each level can grow exponentially. BFS needs to store all nodes at the current level in its queue (and all visited nodes in its set) to ensure optimality. This can quickly exhaust available memory even before finding a solution, making it unfeasible for problems with very large or infinite state spaces.